File: frontend/apps/ios/Pantopus/Features/Businesses/Pages/BusinessPagesViewModel.swift
//
//  BusinessPagesViewModel.swift
//  Pantopus
//
//  C4 — the custom-pages CMS index for a business. Mirrors RN
//  `src/components/business/tabs/PagesTab.tsx:30-108`: create page, delete
//  page, revision history, restore revision. Each row opens the block
//  builder (`BusinessPageBlocksView`).
//

import Foundation
import Observation

/// One row in the pages list.
public struct BusinessPageRow: Sendable, Hashable, Identifiable {
    public let id: String
    public let title: String
    public let slug: String
    public let isDefault: Bool
    public let publishedRevision: Int

    public init(dto: BusinessPageDTO) {
        id = dto.id
        title = dto.title
        slug = dto.slug
        isDefault = dto.isDefault == true
        publishedRevision = dto.publishedRevision ?? 0
    }

    /// `v3` once published, otherwise "Unpublished" — RN's badge copy.
    public var statusLabel: String {
        publishedRevision > 0 ? "v\(publishedRevision)" : "Unpublished"
    }

    public var isPublished: Bool {
        publishedRevision > 0
    }
}

/// One revision-history entry.
public struct BusinessPageRevisionRow: Sendable, Hashable, Identifiable {
    public let id: String
    public let revision: Int
    public let title: String
    public let subtitle: String

    public init(dto: BusinessPageRevisionDTO) {
        id = dto.id
        revision = dto.revision
        let notes = dto.notes?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        title = notes.isEmpty ? "v\(dto.revision)" : "v\(dto.revision) — \(notes)"
        let publisher = dto.publisher?.name
            ?? dto.publisher?.username
            ?? "Unknown"
        let date = BusinessPageRevisionRow.formatted(dto.publishedAt)
        subtitle = date.isEmpty ? publisher : "\(publisher) · \(date)"
    }

    /// Supabase emits both `…Z` and `…​.123456Z`, so try the fractional
    /// parser first and fall back to the plain one.
    private static func formatted(_ iso: String?) -> String {
        guard let iso else { return "" }
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let plain = ISO8601DateFormatter()
        guard let date = fractional.date(from: iso) ?? plain.date(from: iso) else { return "" }
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .none
        return formatter.string(from: date)
    }
}

/// Render state for the pages list.
public enum BusinessPagesState: Sendable, Equatable {
    case loading
    case loaded(rows: [BusinessPageRow])
    case empty
    case error(message: String)
}

@Observable
@MainActor
public final class BusinessPagesViewModel {
    public private(set) var state: BusinessPagesState = .loading
    public private(set) var isCreating = false
    public private(set) var isLoadingRevisions = false
    public private(set) var revisions: [BusinessPageRevisionRow] = []
    /// Page id whose revision history is expanded, nil when collapsed.
    public private(set) var expandedRevisionsPageId: String?
    public private(set) var restoringRevision: Int?

    /// Add-page form.
    public var showsAddForm = false
    public var draftTitle = ""
    public var draftSlug = ""
    public var toastMessage: String?
    /// Page queued for deletion — drives the confirm dialog.
    public var pendingDelete: BusinessPageRow?

    private let businessId: String
    private let api: APIClient

    init(businessId: String, api: APIClient = .shared) {
        self.businessId = businessId
        self.api = api
    }

    // MARK: - Load

    public func load() async {
        if case .loaded = state { return }
        await fetch(showLoading: true)
    }

    public func refresh() async {
        await fetch(showLoading: false)
    }

    private func fetch(showLoading: Bool) async {
        if showLoading { state = .loading }
        do {
            let response: BusinessPagesResponse = try await api.request(
                BusinessPagesEndpoints.pages(businessId: businessId)
            )
            let rows = response.pages.map(BusinessPageRow.init(dto:))
            state = rows.isEmpty ? .empty : .loaded(rows: rows)
        } catch {
            state = .error(message: message(for: error, fallback: "Failed to load pages"))
        }
    }

    // MARK: - Create

    /// Slugs are lowercase `[a-z0-9-]` — RN sanitises as you type.
    public func setSlug(_ raw: String) {
        draftSlug = raw.lowercased().filter { $0.isLowercase || $0.isNumber || $0 == "-" }
    }

    public func createPage() async {
        let title = draftTitle.trimmingCharacters(in: .whitespacesAndNewlines)
        let slug = draftSlug.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !title.isEmpty, !slug.isEmpty else {
            toastMessage = "Title and slug are required"
            return
        }
        guard !isCreating else { return }
        isCreating = true
        defer { isCreating = false }
        do {
            _ = try await api.request(
                BusinessPagesEndpoints.createPage(
                    businessId: businessId,
                    body: CreateBusinessPageRequest(slug: slug, title: title, showInNav: true)
                ),
                as: BusinessPageEnvelope.self
            )
            draftTitle = ""
            draftSlug = ""
            showsAddForm = false
            await fetch(showLoading: false)
        } catch {
            toastMessage = message(for: error, fallback: "Failed to create page")
        }
    }

    // MARK: - Delete

    public func deletePage(_ row: BusinessPageRow) async {
        do {
            _ = try await api.request(
                BusinessPagesEndpoints.deletePage(businessId: businessId, pageId: row.id),
                as: EmptyResponse.self
            )
            if expandedRevisionsPageId == row.id { collapseRevisions() }
            await fetch(showLoading: false)
        } catch {
            toastMessage = message(for: error, fallback: "Failed to delete")
        }
    }

    // MARK: - Revisions

    /// Second tap on the same page collapses the panel — RN's toggle.
    public func toggleRevisions(for row: BusinessPageRow) async {
        if expandedRevisionsPageId == row.id {
            collapseRevisions()
            return
        }
        expandedRevisionsPageId = row.id
        revisions = []
        isLoadingRevisions = true
        defer { isLoadingRevisions = false }
        do {
            let response: BusinessPageRevisionsResponse = try await api.request(
                BusinessPagesEndpoints.revisions(businessId: businessId, pageId: row.id)
            )
            revisions = response.revisions.map(BusinessPageRevisionRow.init(dto:))
        } catch {
            revisions = []
            toastMessage = message(for: error, fallback: "Failed to load revisions")
        }
    }

    public func collapseRevisions() {
        expandedRevisionsPageId = nil
        revisions = []
    }

    public func restore(revision: Int, pageId: String) async {
        guard restoringRevision == nil else { return }
        restoringRevision = revision
        defer { restoringRevision = nil }
        do {
            _ = try await api.request(
                BusinessPagesEndpoints.restoreRevision(
                    businessId: businessId,
                    pageId: pageId,
                    revision: revision
                ),
                as: RestoreBusinessPageRevisionResponse.self
            )
            toastMessage = "Revision v\(revision) restored to draft"
            collapseRevisions()
            await fetch(showLoading: false)
        } catch {
            toastMessage = message(for: error, fallback: "Failed to restore")
        }
    }

    private func message(for error: Error, fallback: String) -> String {
        (error as? APIError)?.errorDescription ?? fallback
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/businesses/pages/BusinessPagesViewModel.kt
@file:Suppress("PackageNaming", "TooManyFunctions")

package app.pantopus.android.ui.screens.businesses.pages

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.business_pages.BusinessPageDto
import app.pantopus.android.data.api.models.business_pages.BusinessPageRevisionDto
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.businesses.BusinessPagesRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle
import java.util.Locale
import javax.inject.Inject

/** Nav-arg key for the business UUID. */
const val BUSINESS_PAGES_BUSINESS_ID_KEY = "businessId"

/** One row in the pages list. */
data class BusinessPageRow(
    val id: String,
    val title: String,
    val slug: String,
    val isDefault: Boolean,
    val publishedRevision: Int,
) {
    val isPublished: Boolean get() = publishedRevision > 0

    /** `v3` once published, otherwise "Unpublished" — RN's badge copy. */
    val statusLabel: String get() = if (isPublished) "v$publishedRevision" else "Unpublished"

    companion object {
        fun from(dto: BusinessPageDto): BusinessPageRow =
            BusinessPageRow(
                id = dto.id,
                title = dto.title,
                slug = dto.slug,
                isDefault = dto.isDefault == true,
                publishedRevision = dto.publishedRevision ?: 0,
            )
    }
}

/** One revision-history entry. */
data class BusinessPageRevisionRow(
    val id: String,
    val revision: Int,
    val title: String,
    val subtitle: String,
) {
    companion object {
        fun from(dto: BusinessPageRevisionDto): BusinessPageRevisionRow {
            val notes = dto.notes?.trim().orEmpty()
            val publisher = dto.publisher?.name ?: dto.publisher?.username ?: "Unknown"
            val date = formatted(dto.publishedAt)
            return BusinessPageRevisionRow(
                id = dto.id,
                revision = dto.revision,
                title = if (notes.isEmpty()) "v${dto.revision}" else "v${dto.revision} — $notes",
                subtitle = if (date.isEmpty()) publisher else "$publisher · $date",
            )
        }

        private fun formatted(iso: String?): String {
            if (iso.isNullOrBlank()) return ""
            return runCatching {
                val instant = Instant.parse(iso)
                DateTimeFormatter
                    .ofLocalizedDate(FormatStyle.MEDIUM)
                    .withLocale(Locale.getDefault())
                    .withZone(ZoneId.systemDefault())
                    .format(instant)
            }.getOrDefault("")
        }
    }
}

/** Render state for the pages list. */
sealed interface BusinessPagesUiState {
    data object Loading : BusinessPagesUiState

    data object Empty : BusinessPagesUiState

    data class Loaded(
        val rows: List<BusinessPageRow>,
    ) : BusinessPagesUiState

    data class Error(
        val message: String,
    ) : BusinessPagesUiState
}

/**
 * C4 — the custom-pages CMS index for a business. Mirrors RN
 * `src/components/business/tabs/PagesTab.tsx:30-108` and iOS
 * `BusinessPagesViewModel`: create page, delete page, revision history,
 * restore revision. Each row opens the block builder.
 */
@HiltViewModel
class BusinessPagesViewModel
    @Inject
    constructor(
        savedStateHandle: SavedStateHandle,
        private val repository: BusinessPagesRepository,
    ) : ViewModel() {
        private val businessId: String = savedStateHandle.get<String>(BUSINESS_PAGES_BUSINESS_ID_KEY).orEmpty()

        private val _state = MutableStateFlow<BusinessPagesUiState>(BusinessPagesUiState.Loading)
        val state: StateFlow<BusinessPagesUiState> = _state.asStateFlow()

        private val _showsAddForm = MutableStateFlow(false)
        val showsAddForm: StateFlow<Boolean> = _showsAddForm.asStateFlow()

        private val _draftTitle = MutableStateFlow("")
        val draftTitle: StateFlow<String> = _draftTitle.asStateFlow()

        private val _draftSlug = MutableStateFlow("")
        val draftSlug: StateFlow<String> = _draftSlug.asStateFlow()

        private val _isCreating = MutableStateFlow(false)
        val isCreating: StateFlow<Boolean> = _isCreating.asStateFlow()

        private val _expandedRevisionsPageId = MutableStateFlow<String?>(null)
        val expandedRevisionsPageId: StateFlow<String?> = _expandedRevisionsPageId.asStateFlow()

        private val _revisions = MutableStateFlow<List<BusinessPageRevisionRow>>(emptyList())
        val revisions: StateFlow<List<BusinessPageRevisionRow>> = _revisions.asStateFlow()

        private val _isLoadingRevisions = MutableStateFlow(false)
        val isLoadingRevisions: StateFlow<Boolean> = _isLoadingRevisions.asStateFlow()

        private val _restoringRevision = MutableStateFlow<Int?>(null)
        val restoringRevision: StateFlow<Int?> = _restoringRevision.asStateFlow()

        private val _toast = MutableStateFlow<String?>(null)
        val toast: StateFlow<String?> = _toast.asStateFlow()

        private var loaded = false

        fun load() {
            if (loaded) return
            fetch(showLoading = true)
        }

        fun refresh() = fetch(showLoading = false)

        private fun fetch(showLoading: Boolean) {
            if (showLoading) _state.value = BusinessPagesUiState.Loading
            viewModelScope.launch {
                when (val result = repository.pages(businessId)) {
                    is NetworkResult.Success -> {
                        val rows = result.data.pages.map(BusinessPageRow::from)
                        loaded = true
                        _state.value =
                            if (rows.isEmpty()) BusinessPagesUiState.Empty else BusinessPagesUiState.Loaded(rows)
                    }
                    is NetworkResult.Failure -> _state.value = BusinessPagesUiState.Error(result.error.message)
                }
            }
        }

        fun toggleAddForm() {
            _showsAddForm.value = !_showsAddForm.value
        }

        fun setTitle(value: String) {
            _draftTitle.value = value
        }

        /** Slugs are lowercase `[a-z0-9-]` — RN sanitises as you type. */
        fun setSlug(raw: String) {
            _draftSlug.value = raw.lowercase().filter { it.isDigit() || it in 'a'..'z' || it == '-' }
        }

        fun consumeToast() {
            _toast.value = null
        }

        fun createPage() {
            val title = _draftTitle.value.trim()
            val slug = _draftSlug.value.trim()
            if (title.isEmpty() || slug.isEmpty()) {
                _toast.value = "Title and slug are required"
                return
            }
            if (_isCreating.value) return
            _isCreating.value = true
            viewModelScope.launch {
                when (val result = repository.createPage(businessId, slug = slug, title = title)) {
                    is NetworkResult.Success -> {
                        _draftTitle.value = ""
                        _draftSlug.value = ""
                        _showsAddForm.value = false
                        fetch(showLoading = false)
                    }
                    is NetworkResult.Failure -> _toast.value = result.error.message
                }
                _isCreating.value = false
            }
        }

        fun deletePage(row: BusinessPageRow) {
            viewModelScope.launch {
                when (val result = repository.deletePage(businessId, row.id)) {
                    is NetworkResult.Success -> {
                        if (_expandedRevisionsPageId.value == row.id) collapseRevisions()
                        fetch(showLoading = false)
                    }
                    is NetworkResult.Failure -> _toast.value = result.error.message
                }
            }
        }

        /** Second tap on the same page collapses the panel — RN's toggle. */
        fun toggleRevisions(row: BusinessPageRow) {
            if (_expandedRevisionsPageId.value == row.id) {
                collapseRevisions()
                return
            }
            _expandedRevisionsPageId.value = row.id
            _revisions.value = emptyList()
            _isLoadingRevisions.value = true
            viewModelScope.launch {
                when (val result = repository.revisions(businessId, row.id)) {
                    is NetworkResult.Success ->
                        _revisions.value = result.data.revisions.map(BusinessPageRevisionRow::from)
                    is NetworkResult.Failure -> {
                        _revisions.value = emptyList()
                        _toast.value = result.error.message
                    }
                }
                _isLoadingRevisions.value = false
            }
        }

        fun collapseRevisions() {
            _expandedRevisionsPageId.value = null
            _revisions.value = emptyList()
        }

        fun restore(
            pageId: String,
            revision: Int,
        ) {
            if (_restoringRevision.value != null) return
            _restoringRevision.value = revision
            viewModelScope.launch {
                when (val result = repository.restoreRevision(businessId, pageId, revision)) {
                    is NetworkResult.Success -> {
                        _toast.value = "Revision v$revision restored to draft"
                        collapseRevisions()
                        fetch(showLoading = false)
                    }
                    is NetworkResult.Failure -> _toast.value = result.error.message
                }
                _restoringRevision.value = null
            }
        }
    }
